File: pipeline/methodtree.py
"""Deterministic, bounded JSON routing trees for browser method assets."""

from __future__ import annotations

import gzip
import hashlib
import io
import json
import re
import unicodedata
from collections import Counter
from dataclasses import dataclass
from pathlib import Path

from files import atomic_write_bytes


GENERATOR = "methods-browser-1"
STATUS = "corpus-extracted-candidates"
NORMALIZATION = "nfkc-lower-alnum-space-1"
MIN_QUERY = 3
SEARCH_RAW = 256 * 1024
SEARCH_GZIP = 64 * 1024
DETAIL_RAW = 128 * 1024
DETAIL_GZIP = 32 * 1024
ROUTER_RAW = 64 * 1024
ROUTER_GZIP = 16 * 1024
MAX_SEARCH_PREFIX = 12
MAX_HASH_PREFIX = 8
HEX = re.compile(r"^[0-9a-f]{64}$")
MAX_PACKAGE_BYTES = 100 * 1024 * 1024


class PackageTooLarge(ValueError):
    """Signal that a browser tier crossed its hard same-origin byte budget."""


@dataclass(frozen=True)
class Limits:
    """Raw and deterministic-gzip byte ceilings for generated JSON."""

    search_raw: int = SEARCH_RAW
    search_gzip: int = SEARCH_GZIP
    detail_raw: int = DETAIL_RAW
    detail_gzip: int = DETAIL_GZIP
    router_raw: int = ROUTER_RAW
    router_gzip: int = ROUTER_GZIP


def json_bytes(value: object) -> bytes:
    """Encode canonical, browser-readable JSON with one trailing newline."""
    return (
        json.dumps(
            value,
            ensure_ascii=False,
            sort_keys=True,
            separators=(",", ":"),
        )
        + "\n"
    ).encode("utf-8")


def gzip_length(content: bytes) -> int:
    """Measure a stable gzip representation for the transfer-size budget."""
    output = io.BytesIO()
    with gzip.GzipFile(fileobj=output, mode="wb", filename="", mtime=0) as stream:
        stream.write(content)
    return len(output.getvalue())


def fits(content: bytes, raw: int, compressed: int) -> bool:
    """Return whether both browser transfer ceilings are satisfied."""
    return len(content) <= raw and gzip_length(content) <= compressed


def compact_row(row: dict) -> dict:
    """Strip evidence while retaining the complete search/result summary."""
    keys = (
        "id",
        "status",
        "label",
        "kind",
        "head",
        "support_count",
        "mention_count",
        "first_year",
        "last_year",
        "scope_counts",
    )
    return {key: row[key] for key in keys}


def row_key(row: dict) -> tuple:
    """Return the canonical extraction ranking with an identity tie-breaker."""
    return (-row["support_count"], row["label"], row["head"], row["id"])


def search_words(label: str) -> tuple[str, ...]:
    """Normalize every searchable label word with the public browser policy."""
    normalized = unicodedata.normalize("NFKC", label).lower()
    normalized = "".join(
        char if char.isascii() and char.isalnum() else " " for char in normalized
    )
    return tuple(
        sorted({word for word in normalized.split() if len(word) >= MIN_QUERY})
    )


def support_bins(rows: list[dict], minimum: int) -> list[dict]:
    """Build deterministic, exhaustive support-frequency histogram bins."""
    edges = [minimum, *(edge for edge in (10, 100, 1000, 10000) if edge > minimum)]
    result = []
    for index, lower in enumerate(edges):
        upper = edges[index + 1] - 1 if index + 1 < len(edges) else None
        result.append(
            {
                "minimum": lower,
                "maximum": upper,
                "count": sum(
                    1
                    for row in rows
                    if row["support_count"] >= lower
                    and (upper is None or row["support_count"] <= upper)
                ),
            }
        )
    return result


def summary_value(source: dict, rows: list[dict]) -> dict:
    """Aggregate exact small statistics without changing candidate semantics."""
    kinds = Counter(row["kind"] for row in rows)
    scopes = Counter()
    years = Counter(row["first_year"] for row in rows)
    for row in rows:
        scopes.update(row["scope_counts"])
    return {
        "schema_version": 1,
        "corpus_manifest_sha256": source["corpus"]["manifest_sha256"],
        "qualified_candidates": len(rows),
        "distinct_extracted_candidates": source["coverage"][
            "distinct_extracted_candidates"
        ],
        "curated_families": source["curated_families"],
        "by_kind": [
            {"kind": kind, "count": kinds[kind]}
            for kind in ("method-noun", "process-technique")
        ],
        "by_scope": [
            {"scope": scope, "count": scopes[scope]}
            for scope in ("likely", "possible", "outside")
        ],
        "by_support": support_bins(rows, source["extraction"]["minimum_support"]),
        "by_first_year": [
            {"year": year, "count": years[year]} for year in sorted(years)
        ],
    }


def id_digest(row: dict) -> str:
    """Return the validated hexadecimal portion of a method candidate ID."""
    prefix, separator, digest = row["id"].partition(":")
    if prefix != "method-candidate" or separator != ":" or not HEX.fullmatch(digest):
        raise ValueError("Method candidate ID cannot be routed")
    return digest


class Store:
    """Write content-addressed children and return strict public descriptors."""

    def __init__(self, root: Path, max_bytes: int = MAX_PACKAGE_BYTES):
        self.root = root
        self.paths: set[str] = set()
        self.max_bytes = max_bytes
        self.total_bytes = 0

    def write(self, stem: str, value: dict, rows: int) -> dict:
        content = json_bytes(value)
        digest = hashlib.sha256(content).hexdigest()
        name = f"{stem}-{digest[:16]}.json"
        if name in self.paths:
            existing = self.root / name
            if not existing.is_file() or existing.read_bytes() != content:
                raise ValueError("Method asset address collision")
        else:
            if self.total_bytes + len(content) > self.max_bytes:
                raise PackageTooLarge(
                    "Method browser package exceeds the 100 MiB Pages boundary"
                )
            atomic_write_bytes(self.root / name, content)
            self.paths.add(name)
            self.total_bytes += len(content)
        return {
            "path": name,
            "encoding": "json",
            "sha256": digest,
            "bytes": len(content),
            "row_count": rows,
        }


def node_desc(descriptor: dict, kind: str, prefix: str, **extra: str) -> dict:
    """Add deterministic routing metadata to one content descriptor."""
    return {"kind": kind, "prefix": prefix, **extra, **descriptor}


def search_leaf(
    store: Store,
    corpus: str,
    prefix: str,
    postings: list[tuple[dict, str]],
    limits: Limits,
    hash_prefix: str = "",
) -> dict | None:
    """Write one deduplicated compact search leaf when it fits both caps."""
    unique = {row["id"]: row for row, _ in postings}
    rows = [compact_row(row) for row in sorted(unique.values(), key=row_key)]
    body = {
        "schema_version": 1,
        "corpus_manifest_sha256": corpus,
        "prefix": prefix,
        "hash_prefix": hash_prefix,
        "rows": rows,
    }
    content = json_bytes(body)
    if not fits(content, limits.search_raw, limits.search_gzip):
        return None
    stem = f"search-{prefix[:MAX_SEARCH_PREFIX]}"
    descriptor = store.write(stem, body, len(rows))
    extra = {"hash_prefix": hash_prefix} if hash_prefix else {}
    return node_desc(descriptor, "leaf", prefix, **extra)


def search_hash(
    store: Store,
    corpus: str,
    prefix: str,
    postings: list[tuple[dict, str]],
    limits: Limits,
    hash_prefix: str = "",
) -> dict:
    """Split an overfull terminal word bucket by candidate identity hash."""
    leaf = search_leaf(store, corpus, prefix, postings, limits, hash_prefix)
    if leaf is not None:
        if not hash_prefix:
            leaf["route_mode"] = "hash"
        return leaf
    if len(hash_prefix) >= MAX_HASH_PREFIX:
        raise ValueError("Search leaf exceeds its cap at maximum hash depth")
    groups: dict[str, list[tuple[dict, str]]] = {}
    for posting in postings:
        digest = id_digest(posting[0])
        child = digest[: len(hash_prefix) + 1]
        groups.setdefault(child, []).append(posting)
    shards = [
        search_hash(store, corpus, prefix, groups[key], limits, key)
        for key in sorted(groups)
    ]
    body = {
        "schema_version": 1,
        "corpus_manifest_sha256": corpus,
        "route_kind": "search",
        "route_mode": "hash",
        "prefix": prefix,
        "hash_prefix": hash_prefix,
        "row_count": sum(item["row_count"] for item in shards),
        "shards": shards,
    }
    content = json_bytes(body)
    if not fits(content, limits.router_raw, limits.router_gzip):
        raise ValueError("Search hash router exceeds its cap")
    descriptor = store.write("search-route", body, body["row_count"])
    return node_desc(
        descriptor,
        "router",
        prefix,
        route_mode="hash",
        hash_prefix=hash_prefix,
    )


def search_node(
    store: Store,
    corpus: str,
    prefix: str,
    postings: list[tuple[dict, str]],
    limits: Limits,
) -> dict:
    """Build the bounded adaptive word-prefix tree for one posting bucket."""
    if len(prefix) >= MIN_QUERY:
        leaf = search_leaf(store, corpus, prefix, postings, limits)
        if leaf is not None:
            return leaf
    if len(prefix) >= MAX_SEARCH_PREFIX:
        return search_hash(store, corpus, prefix, postings, limits)
    exact: list[tuple[dict, str]] = []
    groups: dict[str, list[tuple[dict, str]]] = {}
    for posting in postings:
        word = posting[1]
        if len(word) == len(prefix):
            exact.append(posting)
        else:
            child = word[: len(prefix) + 1]
            groups.setdefault(child, []).append(posting)
    shards = []
    if exact:
        shards.append(search_hash(store, corpus, prefix, exact, limits))
    shards.extend(
        search_node(store, corpus, key, groups[key], limits) for key in sorted(groups)
    )
    body = {
        "schema_version": 1,
        "corpus_manifest_sha256": corpus,
        "route_kind": "search",
        "route_mode": "word",
        "prefix": prefix,
        "row_count": sum(item["row_count"] for item in shards),
        "shards": shards,
    }
    content = json_bytes(body)
    if not fits(content, limits.router_raw, limits.router_gzip):
        raise ValueError("Search word router exceeds its cap")
    descriptor = store.write("search-route", body, body["row_count"])
    return node_desc(descriptor, "router", prefix, route_mode="word")


def build_search(store: Store, corpus: str, rows: list[dict], limits: Limits) -> dict:
    """Build the lazily loaded search manifest and all adaptive children."""
    postings = [(row, word) for row in rows for word in search_words(row["label"])]
    if any(not search_words(row["label"]) for row in rows):
        raise ValueError("Every method candidate must have a searchable word")
    groups: dict[str, list[tuple[dict, str]]] = {}
    for posting in postings:
        groups.setdefault(posting[1][:MIN_QUERY], []).append(posting)
    shards = [
        search_node(store, corpus, prefix, groups[prefix], limits)
        for prefix in sorted(groups)
    ]
    body = {
        "schema_version": 1,
        "corpus_manifest_sha256": corpus,
        "normalization": NORMALIZATION,
        "minimum_query_length": MIN_QUERY,
        "row_count": sum(item["row_count"] for item in shards),
        "shards": shards,
    }
    content = json_bytes(body)
    if not fits(content, limits.router_raw, limits.router_gzip):
        shards = route_roots(store, corpus, "search", shards, limits)
        body["shards"] = shards
        content = json_bytes(body)
        if not fits(content, limits.router_raw, limits.router_gzip):
            raise ValueError("Search routing manifest exceeds its cap")
    return store.write("search", body, body["row_count"])


def detail_leaf(
    store: Store,
    corpus: str,
    prefix: str,
    rows: list[dict],
    limits: Limits,
) -> dict | None:
    """Write one full candidate leaf when both detail ceilings permit it."""
    ordered = sorted(rows, key=row_key)
    body = {
        "schema_version": 1,
        "corpus_manifest_sha256": corpus,
        "prefix": prefix,
        "rows": ordered,
    }
    content = json_bytes(body)
    if not fits(content, limits.detail_raw, limits.detail_gzip):
        return None
    descriptor = store.write(f"detail-{prefix}", body, len(ordered))
    return node_desc(descriptor, "leaf", prefix)


def detail_node(
    store: Store,
    corpus: str,
    prefix: str,
    rows: list[dict],
    limits: Limits,
) -> dict:
    """Build one adaptive leading-ID-hash detail subtree."""
    if len(prefix) >= 2:
        leaf = detail_leaf(store, corpus, prefix, rows, limits)
        if leaf is not None:
            return leaf
    if len(prefix) >= MAX_HASH_PREFIX:
        raise ValueError("Detail leaf exceeds its cap at maximum hash depth")
    groups: dict[str, list[dict]] = {}
    for row in rows:
        child = id_digest(row)[: len(prefix) + 1]
        groups.setdefault(child, []).append(row)
    shards = [
        detail_node(store, corpus, key, groups[key], limits) for key in sorted(groups)
    ]
    body = {
        "schema_version": 1,
        "corpus_manifest_sha256": corpus,
        "route_kind": "detail",
        "route_mode": "hash",
        "prefix": prefix,
        "row_count": sum(item["row_count"] for item in shards),
        "shards": shards,
    }
    content = json_bytes(body)
    if not fits(content, limits.router_raw, limits.router_gzip):
        raise ValueError("Detail hash router exceeds its cap")
    descriptor = store.write("detail-route", body, body["row_count"])
    return node_desc(descriptor, "router", prefix, route_mode="hash")


def build_details(store: Store, corpus: str, rows: list[dict], limits: Limits) -> dict:
    """Build the lazy detail manifest and full evidence-bearing leaves."""
    groups: dict[str, list[dict]] = {}
    for row in rows:
        groups.setdefault(id_digest(row)[:2], []).append(row)
    shards = [
        detail_node(store, corpus, prefix, groups[prefix], limits)
        for prefix in sorted(groups)
    ]
    body = {
        "schema_version": 1,
        "corpus_manifest_sha256": corpus,
        "prefix_bits": 8,
        "row_count": sum(item["row_count"] for item in shards),
        "shards": shards,
    }
    content = json_bytes(body)
    if not fits(content, limits.router_raw, limits.router_gzip):
        shards = route_roots(store, corpus, "detail", shards, limits)
        body["shards"] = shards
        content = json_bytes(body)
        if not fits(content, limits.router_raw, limits.router_gzip):
            raise ValueError("Detail routing manifest exceeds its cap")
    return store.write("details", body, body["row_count"])


def route_body(
    store: Store,
    corpus: str,
    route_kind: str,
    prefix: str,
    shards: list[dict],
    limits: Limits,
) -> dict:
    """Write a compact hierarchy node, recursively grouping if necessary."""
    mode = "word" if route_kind == "search" else "hash"
    body = {
        "schema_version": 1,
        "corpus_manifest_sha256": corpus,
        "route_kind": route_kind,
        "route_mode": mode,
        "prefix": prefix,
        "row_count": sum(item["row_count"] for item in shards),
        "shards": shards,
    }
    content = json_bytes(body)
    if not fits(content, limits.router_raw, limits.router_gzip):
        shards = route_roots(store, corpus, route_kind, shards, limits, len(prefix))
        body["shards"] = shards
        content = json_bytes(body)
        if not fits(content, limits.router_raw, limits.router_gzip):
            raise ValueError("Method hierarchy router exceeds its cap")
    stem = "search-route" if route_kind == "search" else "detail-route"
    descriptor = store.write(stem, body, body["row_count"])
    return node_desc(descriptor, "router", prefix, route_mode=mode)


def route_roots(
    store: Store,
    corpus: str,
    route_kind: str,
    shards: list[dict],
    limits: Limits,
    depth: int = 0,
) -> list[dict]:
    """Collapse a wide flat route list into bounded deterministic radix nodes."""
    groups: dict[str, list[dict]] = {}
    for shard in shards:
        prefix = shard["prefix"]
        if len(prefix) <= depth:
            raise ValueError("Method route cannot be hierarchically partitioned")
        groups.setdefault(prefix[: depth + 1], []).append(shard)
    result = []
    for prefix in sorted(groups):
        children = groups[prefix]
        if len(children) == 1:
            result.append(children[0])
        else:
            result.append(
                route_body(store, corpus, route_kind, prefix, children, limits)
            )
    return result

File: web/src/lib/methodtree.ts
import { hasOnlyKeys, isRecord, isString, type RecordValue } from "./guards";
import {
  METHOD_CAPS,
  methodCount,
  readMethodRows,
  type MethodAsset,
  type MethodIndex,
  type MethodRow,
} from "./methods";
import { normalizeMethodQuery } from "./methodview";

const DIGEST = /^[0-9a-f]{64}$/;
const SEARCH_LEAF = /^search-[0-9a-z]{3,12}-[0-9a-f]{16}\.json$/;
const DETAIL_LEAF = /^detail-[0-9a-f]{2,8}-[0-9a-f]{16}\.json$/;
const SEARCH_ROUTE = /^search-route-[0-9a-f]{16}\.json$/;
const DETAIL_ROUTE = /^detail-route-[0-9a-f]{16}\.json$/;
const NODE_BASE = ["path", "encoding", "sha256", "bytes", "row_count"];

export type MethodRoute = "search" | "detail";
export type RouteMode = "word" | "hash" | "ordinal";
export type MethodNode = MethodAsset & {
  kind: "leaf" | "router";
  prefix: string;
  routeMode?: RouteMode;
  hashPrefix?: string;
  startOrdinal?: number;
  endOrdinal?: number;
};
export type MethodRoot = { rowCount: number; shards: MethodNode[] };
export type SearchLeaf = { rows?: MethodRow[]; ordinals?: number[] };

function exact(value: RecordValue, keys: ReadonlySet<string>): boolean {
  return hasOnlyKeys(value, keys) && Object.keys(value).length === keys.size;
}

function addressed(path: string, digest: string): boolean {
  return path.endsWith(`-${digest.slice(0, 16)}.json`);
}

function nodeKeys(value: RecordValue): Set<string> {
  return new Set([
    ...NODE_BASE,
    "kind",
    "prefix",
    ...(value.route_mode === undefined ? [] : ["route_mode"]),
    ...(value.hash_prefix === undefined ? [] : ["hash_prefix"]),
    ...(value.start_ordinal === undefined ? [] : ["start_ordinal"]),
    ...(value.end_ordinal === undefined ? [] : ["end_ordinal"]),
  ]);
}

export function readMethodNode(
  value: unknown,
  route: MethodRoute,
  tier: MethodIndex["tier"],
): MethodNode | null {
  if (!isRecord(value) || (value.kind !== "leaf" && value.kind !== "router")) {
    return null;
  }
  const mode = value.route_mode;
  const hash = value.hash_prefix;
  const ordinal = route === "detail" && tier === "catalog-only";
  if (
    !exact(value, nodeKeys(value)) ||
    !isString(value.path) ||
    value.encoding !== "json" ||
    !isString(value.sha256) ||
    !DIGEST.test(value.sha256) ||
    !addressed(value.path, value.sha256) ||
    !methodCount(value.bytes) ||
    value.bytes < 2 ||
    !methodCount(value.row_count) ||
    value.row_count < 1 ||
    !isString(value.prefix)
  ) {
    return null;
  }
  if (ordinal) {
    if (
      mode !== "ordinal" ||
      !/^[0-9a-f]{8}$/.test(value.prefix) ||
      !methodCount(value.start_ordinal) ||
      !methodCount(value.end_ordinal) ||
      value.end_ordinal < value.start_ordinal ||
      value.end_ordinal - value.start_ordinal + 1 !== value.row_count ||
      hash !== undefined
    ) {
      return null;
    }
  } else {
    if (value.start_ordinal !== undefined || value.end_ordinal !== undefined)
      return null;
    if (value.kind === "router" && mode !== "word" && mode !== "hash") return null;
    if (value.kind === "leaf" && mode !== undefined && mode !== "hash") return null;
    if (route === "detail" && (mode === "word" || hash !== undefined)) return null;
  }
  if (route === "search") {
    if (!/^[a-z0-9]{0,12}$/.test(value.prefix)) return null;
    if (hash !== undefined && (!isString(hash) || !/^[0-9a-f]{0,8}$/.test(hash))) {
      return null;
    }
    if (value.kind === "router" && mode === "hash" && hash === undefined) return null;
    if (mode === "word" && hash !== undefined) return null;
    if (value.kind === "leaf" && value.prefix.length < 3) return null;
  } else if (!ordinal && !/^[0-9a-f]{0,8}$/.test(value.prefix)) {
    return null;
  }
  const pattern =
    value.kind === "router"
      ? route === "search"
        ? SEARCH_ROUTE
        : DETAIL_ROUTE
      : route === "search"
        ? SEARCH_LEAF
        : DETAIL_LEAF;
  const cap =
    value.kind === "router"
      ? METHOD_CAPS.router
      : route === "search"
        ? METHOD_CAPS.search
        : METHOD_CAPS.detail;
  if (!pattern.test(value.path) || value.bytes > cap) return null;
  return {
    path: value.path,
    encoding: "json",
    sha256: value.sha256,
    bytes: value.bytes,
    rowCount: value.row_count,
    kind: value.kind,
    prefix: value.prefix,
    routeMode: mode as RouteMode | undefined,
    hashPrefix: hash as string | undefined,
    startOrdinal: value.start_ordinal as number | undefined,
    endOrdinal: value.end_ordinal as number | undefined,
  };
}

export function readMethodNodes(
  values: unknown,
  route: MethodRoute,
  index: MethodIndex,
): MethodNode[] | null {
  if (!Array.isArray(values) || values.length < 1) return null;
  const nodes = values.map((value) => readMethodNode(value, route, index.tier));
  if (nodes.some((node) => node === null)) return null;
  const valid = nodes as MethodNode[];
  if (new Set(valid.map((node) => node.path)).size !== valid.length) return null;
  const keys = valid.map((node) =>
    [node.prefix, node.hashPrefix ?? "", node.startOrdinal ?? -1, node.path].join("\0"),
  );
  return keys.join() === [...keys].sort().join() ? valid : null;
}

export function nodeRows(nodes: readonly MethodNode[]): number {
  return nodes.reduce((total, node) => total + node.rowCount, 0);
}

export function readSearchRoot(value: unknown, index: MethodIndex): MethodRoot | null {
  const catalog = index.tier === "catalog-only";
  const keys = new Set([
    "schema_version",
    "corpus_manifest_sha256",
    ...(catalog ? ["full_asset_sha256"] : []),
    "normalization",
    "minimum_query_length",
    "row_count",
    "shards",
  ]);
  if (
    !isRecord(value) ||
    !exact(value, keys) ||
    value.schema_version !== 1 ||
    value.corpus_manifest_sha256 !== index.corpusDigest ||
    (catalog && value.full_asset_sha256 !== index.download.sha256) ||
    value.normalization !== "nfkc-lower-alnum-space-1" ||
    value.minimum_query_length !== 3 ||
    value.row_count !== index.search.rowCount
  ) {
    return null;
  }
  const shards = readMethodNodes(value.shards, "search", index);
  return shards && nodeRows(shards) === value.row_count
    ? { rowCount: value.row_count, shards }
    : null;
}

export function readDetailRoot(value: unknown, index: MethodIndex): MethodRoot | null {
  const catalog = index.tier === "catalog-only";
  const keys = new Set(
    catalog
      ? [
          "schema_version",
          "corpus_manifest_sha256",
          "full_asset_sha256",
          "route_kind",
          "route_mode",
          "start_ordinal",
          "end_ordinal",
          "row_count",
          "shards",
        ]
      : [
          "schema_version",
          "corpus_manifest_sha256",
          "prefix_bits",
          "row_count",
          "shards",
        ],
  );
  if (
    !isRecord(value) ||
    !exact(value, keys) ||
    value.schema_version !== 1 ||
    value.corpus_manifest_sha256 !== index.corpusDigest ||
    value.row_count !== index.details.rowCount
  ) {
    return null;
  }
  if (
    catalog
      ? value.full_asset_sha256 !== index.download.sha256 ||
        value.route_kind !== "detail" ||
        value.route_mode !== "ordinal" ||
        value.start_ordinal !== 0 ||
        value.end_ordinal !== index.qualifiedCandidates - 1
      : value.prefix_bits !== 8
  ) {
    return null;
  }
  const shards = readMethodNodes(value.shards, "detail", index);
  return shards && nodeRows(shards) === value.row_count
    ? { rowCount: value.row_count, shards }
    : null;
}

export function readRouter(
  value: unknown,
  index: MethodIndex,
  node: MethodNode,
  route: MethodRoute,
): MethodNode[] | null {
  const catalog = index.tier === "catalog-only";
  const ordinal = route === "detail" && catalog;
  const hash = route === "search" && node.routeMode === "hash";
  const keys = new Set([
    "schema_version",
    "corpus_manifest_sha256",
    ...(catalog ? ["full_asset_sha256"] : []),
    "route_kind",
    "route_mode",
    "prefix",
    ...(ordinal ? ["start_ordinal", "end_ordinal"] : []),
    "row_count",
    "shards",
    ...(hash ? ["hash_prefix"] : []),
  ]);
  if (
    !isRecord(value) ||
    !exact(value, keys) ||
    value.schema_version !== 1 ||
    value.corpus_manifest_sha256 !== index.corpusDigest ||
    (catalog && value.full_asset_sha256 !== index.download.sha256) ||
    value.route_kind !== route ||
    value.route_mode !== node.routeMode ||
    value.prefix !== node.prefix ||
    value.row_count !== node.rowCount ||
    (hash && value.hash_prefix !== node.hashPrefix) ||
    (ordinal &&
      (value.start_ordinal !== node.startOrdinal ||
        value.end_ordinal !== node.endOrdinal))
  ) {
    return null;
  }
  const shards = readMethodNodes(value.shards, route, index);
  if (!shards || nodeRows(shards) !== node.rowCount) return null;
  const routes = shards.map((child) =>
    [child.prefix, child.hashPrefix ?? "", child.startOrdinal ?? -1].join("\0"),
  );
  if (new Set(routes).size !== routes.length) return null;
  if (ordinal) {
    let next = node.startOrdinal;
    for (const child of shards) {
      if (child.startOrdinal !== next) return null;
      next = (child.endOrdinal ?? -1) + 1;
    }
    if (next !== (node.endOrdinal ?? -1) + 1) return null;
  } else if (
    route === "detail" &&
    shards.some(
      (child) =>
        !child.prefix.startsWith(node.prefix) ||
        child.prefix.length <= node.prefix.length,
    )
  ) {
    return null;
  } else if (
    route === "search" &&
    node.routeMode === "word" &&
    shards.some((child) => {
      const terminal = child.routeMode === "hash" || child.hashPrefix !== undefined;
      return terminal
        ? child.prefix !== node.prefix
        : !child.prefix.startsWith(node.prefix) ||
            child.prefix.length <= node.prefix.length;
    })
  ) {
    return null;
  } else if (
    route === "search" &&
    node.routeMode === "hash" &&
    shards.some(
      (child) =>
        child.prefix !== node.prefix ||
        !child.hashPrefix?.startsWith(node.hashPrefix ?? "") ||
        child.hashPrefix.length <= (node.hashPrefix ?? "").length,
    )
  ) {
    return null;
  }
  return shards;
}

export function readSearchLeaf(
  value: unknown,
  index: MethodIndex,
  node: MethodNode,
): SearchLeaf | null {
  const catalog = index.tier === "catalog-only";
  const keys = new Set([
    "schema_version",
    "corpus_manifest_sha256",
    ...(catalog ? ["full_asset_sha256"] : []),
    "prefix",
    "hash_prefix",
    catalog ? "ordinals" : "rows",
  ]);
  if (
    !isRecord(value) ||
    !exact(value, keys) ||
    value.schema_version !== 1 ||
    value.corpus_manifest_sha256 !== index.corpusDigest ||
    (catalog && value.full_asset_sha256 !== index.download.sha256) ||
    value.prefix !== node.prefix ||
    value.hash_prefix !== (node.hashPrefix ?? "")
  ) {
    return null;
  }
  if (catalog) {
    if (!Array.isArray(value.ordinals) || value.ordinals.length !== node.rowCount) {
      return null;
    }
    const ordinals = value.ordinals;
    if (
      ordinals.some(
        (item, position) =>
          !methodCount(item) ||
          item >= index.qualifiedCandidates ||
          (position > 0 && ordinals[position - 1] >= item),
      )
    ) {
      return null;
    }
    return { ordinals: ordinals as number[] };
  }
  const rows = readMethodRows(value.rows, index.minimumSupport, node.rowCount);
  if (
    !rows ||
    rows.some(
      (row) =>
        normalizeMethodQuery(row.label)
          .split(" ")
          .every((word) => !word.startsWith(node.prefix)) ||
        (node.hashPrefix !== undefined &&
          !row.id.slice("method-candidate:".length).startsWith(node.hashPrefix)),
    )
  ) {
    return null;
  }
  return { rows };
}
